File: leetcode/kthSmallestinBST/ kthSmallestinBST.java
//recursive
int count = 0;
int result = Integer.MIN_VALUE;

public int kthSmallest(TreeNode root, int k) {
    traverse(root, k);
    return result;
}

public void traverse(TreeNode root, int k) {
    if(root == null) return;
    traverse(root.left, k);
    count ++;
    if(count == k) result = root.val;
    traverse(root.right, k);       
}
//iterative
 public int kthSmallest(TreeNode root, int k) {
     Stack<TreeNode> stack = new Stack<TreeNode>();
     TreeNode p = root;
     int count = 0;

     while(!stack.isEmpty() || p != null) {
         if(p != null) {
             stack.push(p);  // Just like recursion
             p = p.left;   

         } else {
            TreeNode node = stack.pop();
            if(++count == k) return node.val; 
            p = node.right;
         }
     }

     return Integer.MIN_VALUE;
 }

File: leetcode/kthSmallestinBST/kthSmallestinBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int retVal = 0;
        kthSmallest_helper(root,k,retVal);
        return retVal;
    }
    
    void kthSmallest_helper(TreeNode* root, int &k, int &retVal){
        if(!root||k<=0)
        return ;
        //first traverse to the bottom left
        kthSmallest_helper(root->left,k,retVal);
        k--;
        if(k==0){
            retVal = root->val;
            return;
        }
        kthSmallest_helper(root->right,k,retVal);
    }
};
